# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd
df1 = pd.read_csv('D:\pytho练习\pythonProject3\e_commerce.csv')
df1.head()

<>:2: SyntaxWarning: invalid escape sequence '\p'
<>:2: SyntaxWarning: invalid escape sequence '\p'
C:\Users\林涵\AppData\Local\Temp\ipykernel_17380\176243545.py:2: SyntaxWarning: invalid escape sequence '\p'
  df1 = pd.read_csv('D:\pytho练习\pythonProject3\e_commerce.csv')


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

### 评估整齐度

In [2]:
df1.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
162624    550512    85199L  LARGE HANGING IVORY & RED WOOD BIRD         1   
384334    570116     23371      SET 36 COLOUR PENCILS SPACEBOY          3   
517460    580058     22910    PAPER CHAIN KIT VINTAGE CHRISTMAS         6   
54519     540954     20752      BLUE POLKADOT WASHING UP GLOVES         1   
102430    545020     22697      GREEN REGENCY TEACUP AND SAUCER         6   
450790    575177     22065       CHRISTMAS PUDDING TRINKET POT         10   
306250    563745     22610             PENS ASSORTED FUNNY FACE        72   
13326     537434     21210   SET OF 72 RETROSPOT PAPER  DOILIES         1   
143916    548714     21080  SET/20 RED RETROSPOT PAPER NAPKINS          2   
504687    578977    90186B     CRYSTAL HOOP EARRING FLORAL LEAF         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
162624   4/18/2011 16:31       0.65     18069.0  United Kingdom  
384334   10/7/2011 12:38       1.25     12856.0  United Kingdom  
517460   12/1/2011 13:19       2.95     14367.0  United Kingdom  
54519    1/12/2011 13:19       2.10     14606.0  United Kingdom  
102430   2/25/2011 14:03       2.95     13792.0  United Kingdom  
450790   11/8/2011 18:41       1.63         NaN  United Kingdom  
306250    8/19/2011 9:38       0.19     15159.0  United Kingdom  
13326    12/6/2010 16:57       2.98         NaN  United Kingdom  
143916    4/3/2011 15:07       0.85     17337.0  United Kingdom  
504687  11/27/2011 15:18       2.95     17231.0  United Kingdom

从抽样10行数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易的各个变量，因此不存在结构问题。

### 评估数据干净度

In [3]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共541909行，而Description、CustomerID变量存在缺失值。
此外，InvoiceDate的数据类型应为日期，CustomerID的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失数据

In [4]:
df1[df1["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

一共有1454条交易数据缺失Description变量值
从输出结果来看，这些缺失Description的交易数据，UnitPrice都为0，为了验证猜想，我们增加筛选条件。

In [5]:
df1[(df1["Description"].isnull()) & (df1["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来的结果为0条，说明缺失Description值的数据，同时也不具备有效的UnitPrice值。
产品名称和产品单价同时缺失，我认为数据无法提供有效含义，因此后续可以删除。

In [6]:
df1[df1["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

客户编号不是分析畅销商品的必要变量，而且虽然客户编号缺失，但交易数据依旧有效，所以保留此观察值。

#### 评估重复数据

虽然交易编码，发票编码，客户编码是唯一标识，但一次交易可以包括多个商品，可以有多此重复情况；因此不评估重复数据。

#### 评估不一致数据

In [7]:
df1["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从变量值看，“USA”、“United States”均表示美国，“United Kingdom”、“UK”、“U.K.”均在表示英国，因此应该统一，保留一个指代值

#### 评估无效和错误数据
通过DateFrame的describe方法，来对数据统计进行快速了解

In [8]:
df1.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [9]:
df1[df1["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从结果来看，为负数的观察值，发票编码均以C开头，是取消单，为验证猜想，增加筛选条件。

In [10]:
df1[(df1["Quantity"] < 0) & (df1["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

猜想错误，根据以上结果单价均为0，因此增加条件

In [11]:
df1[(df1["Quantity"] < 0) & (df1["InvoiceNo"].str[0] != "C") & (df1["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

猜想正确，Quantity为0的情况，要么交易被取消了，要么单价为0，所以在后续可以删除。
接下来筛选出UnitPrice为负数的行

In [12]:
df1[df1["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

为坏账调整，后续删除

## 清理数据

根据前面的评估得到的结论，我们要进行的数据清理：
- 把'InvoiceDate'变量的数据类型改为日期时间
- 把'CustomerID'变量的数据类型改为字符串
- 把'Description'变量缺失的观察值删除
- 把'Country'变量'USA'替换为'United States'
- 把'Country'变量'UK'、'U.K.'替换为'United Kingdom'
- 把'Quantity'变量为负数的观察值删除
- 把'UnitPrice'变量为负数的观察值删除

In [13]:
df2 = df1.copy()
df2.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

把'InvoiceDate'变量的数据类型改为日期时间

In [14]:
df2["InvoiceDate"] = pd.to_datetime(df2["InvoiceDate"])
df2["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把'CustomerID'变量的数据类型改为字符串

In [15]:
df2["CustomerID"] = df2["CustomerID"].astype("str")
df2["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

结尾.0删除

In [16]:
df2["CustomerID"] = df2["CustomerID"].str.slice(start=0, stop=-2)
df2["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把'Description'变量缺失的观察值删除

In [17]:
df2 = df2.dropna(subset=["Description"])

In [18]:
df2["Description"].isnull().sum()

0

把’Country’变量’USA’替换为’United States’

In [19]:
df2["Country"] = df2["Country"].replace({"USA":"United States"})

In [20]:
len(df2[df2["Country"] =="USA"])

0

把'Country'变量'UK'、'U.K.'替换为'United Kingdom'

In [21]:
df2["Country"] = df2["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})

In [22]:
len(df2[df2["Country"] =="UK"])

0

In [23]:
len(df2[df2["Country"] =="U.K."])

0

把'Quantity'变量为负数的观察值删除

In [24]:
df2 = df2[df2["Quantity"] >= 0]

In [25]:
len(df2[df2["Quantity"] < 0])

0

把'UnitPrice'变量为负数的观察值删除

In [26]:
df2 = df2[df2["UnitPrice"] >= 0]

In [27]:
len(df2[df2["UnitPrice"] < 0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv

In [28]:
df2.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom  
5 2010-12-01 08:26:00       7.65      17850  United Kingdom  
6 2010-12-01 08:26:00       4.25      17850  United Kingdom  
7 2010-12-01 08:28:00       1.85      17850  United Kingdom  
8 2010-12-01 08:28:00       1.85      17850  United Kingdom  
9 2010-12-01 08:34:00       1.69      13047  United Kingdom

In [29]:
df2.to_csv("e_commerce_cleaned.csv",index=False)

In [30]:
pd.read_csv("e_commerce_cleaned.csv").head(10)

C:\Users\林涵\AppData\Local\Temp\ipykernel_17380\2059381091.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head(10)


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom  
5  2010-12-01 08:26:00       7.65      17850  United Kingdom  
6  2010-12-01 08:26:00       4.25      17850  United Kingdom  
7  2010-12-01 08:28:00       1.85      17850  United Kingdom  
8  2010-12-01 08:28:00       1.85      17850  United Kingdom  
9  2010-12-01 08:34:00       1.69      13047  United Kingdom